### ISSUE TITLE
Infura not working on Contract Contructor() include eth.send_raw_transaction suggestion?

### ISSUE THREAD

**Issue by Gresliebear** [ethereum/web3.py#2155]
* Version: x.x.x
* Python: 3.9
* OS: Windows 10 
* `pip freeze` output

```
[2021-09-18 23:10:53,398] ERROR in app: Exception on / [GET]
Traceback (most recent call last):
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\flask\app.py", line 2070, in wsgi_app
    response = self.full_dispatch_request()
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\flask\app.py", line 1515, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\flask\app.py", line 1513, in full_dispatch_request
    rv = self.dispatch_request()
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\flask\app.py", line 1499, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "E:\OpenVesselFin\jubilant-market\PythonServer\wsgi.py", line 92, in index
    tx_hash = Greeter.constructor().transact()
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\eth_utils\decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\web3\contract.py", line 663, in transact
    return self.web3.eth.send_transaction(transact_transaction)
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\web3\eth.py", line 686, in send_transaction
    return self._send_transaction(transaction)
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\web3\module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\web3\manager.py", line 187, in request_blocking
    return self.formatted_response(response,
  File "e:\openvesselfin\jubilant-market\pythonserver\vessel_env\lib\site-packages\web3\manager.py", line 168, in formatted_response
    raise ValueError(response["error"])
ValueError: {'code': -32601, 'message': 'The method eth_sendTransaction does not exist/is not available'}```
```

### What was wrong?

I am calling the infura API and would like to use the Contract.constructor().transact() but it uses eth.send_transaction(transact_transaction) which for infurai doesn't work instead I have to use Eth,send_raw_transaction(raw_transaction). I am going to change to local host solution but I don't know how to implement a pull request for including eth.send_raw_transaction to the constructor() decorator

![image](https://user-images.githubusercontent.com/13157730/133914298-da2c37db-4200-4712-b97b-8ef6b3bf764b.png)

### How can it be fixed?

Fill this section in if you know how this could or should be fixed.

I think including the if statement passes the key to check if was signed, or just past the transaction object after it has been signed. or write a new function to check that transaction has been signed or not. I don't know the code base well enough to implement this. However I need to move on in a hackathon so I have to figure out local host. because the only reason I am using Contract Contructor is to access Solidity code through python. 

I know you have no reason to follow through on this but I think this will be essential for multi-key signs so know not depending on a single person to local host or sign transactions off

![image](https://user-images.githubusercontent.com/13157730/133914449-5ff185e9-8e90-4059-8ea5-d38405fdc3da.png)

---
**Note:** We prefer to use issues to track our work. If you think you've encountered a bug in web3py or 
have a feature request, you're in the right place. If you have implementation or usage questions, 
please refer to our [documentation](https://web3py.readthedocs.io/en/latest/) and/or join the conversation 
on [discord](https://discord.gg/GHryRvPB84).

**Comment by CaixiangFan** [comment]
I found this also happens to Besu when trying to send a transaction through deployed smart contract. In my case, the following value error happened when calling the _contract.functions.set(1).transact()_ function in [SimpleStorage](https://github.com/ConsenSys/quorum-dev-quickstart/blob/master/files/common/smart_contracts/contracts/SimpleStorage.sol).

`ValueError: {'code': -32604, 'message': 'The method eth_sendTransaction is not supported. Use eth_sendRawTransaction to send a signed transaction to Besu.'}`

Besu: 21.10
web3: 5.23.0

**Comment by kclowes** [comment]
This isn't a bug, but the docs could be clearer so I'll leave this open. See [here](https://web3py.readthedocs.io/en/stable/examples.html#using-infura-rinkeby-node) for an example of how to build the transaction, sign it, and then send it using `sendRawTransaction` using a node that doesn't handle your private key.

**Comment by Gresliebear** [comment]
I have read that in the docs but the reason I brought this up is that I enjoy the contract functionality where sol compile translate solidity code into python Object calls but this is not possible with Infura and besu API. I have proceeded to set up Geth but its failed to sync and errors. 

Mabye I am using it wrong

**Comment by B3nac** [comment]
So the main action item here is to clear up the documentation about using `eth.send_transaction` and how Infura and Besu are not compatible with this call? I believe only local nodes would be compatible with this function call.

**Comment by kclowes** [comment]
> So the main action item here is to clear up the documentation about using eth.send_transaction and how Infura and Besu are not compatible with this call?

It's been so long that I'm not sure what I was thinking initially, but it would be nice to find a place to highlight that you can add the `construct_sign_and_send_raw_middleware`. Maybe in the troubleshooting section? Docs for the middleware are [here](https://web3py.readthedocs.io/en/stable/middleware.html#web3.middleware.construct_sign_and_send_raw_middleware). I haven't looked at Besu at all, but I  believe that middleware should work on Infura at least.

**Comment by B3nac** [comment]
> > So the main action item here is to clear up the documentation about using eth.send_transaction and how Infura and Besu are not compatible with this call?
> 
> It's been so long that I'm not sure what I was thinking initially, but it would be nice to find a place to highlight that you can add the `construct_sign_and_send_raw_middleware`. Maybe in the troubleshooting section? Docs for the middleware are [here](https://web3py.readthedocs.io/en/stable/middleware.html#web3.middleware.construct_sign_and_send_raw_middleware). I haven't looked at Besu at all, but I believe that middleware should work on Infura at least.

I have an idea, I'll send in a pull request for review!

**Comment by B3nac** [comment]
Hi @kclowes I added an example. Let me know if I should change the wording or anything. In the snippet I used `os.environ.get()` because importing the data from an environment variable is generally more secure than hard coding the private key. Cheers!

**Comment by kclowes** [comment]
Closed via #2410. Thanks @B3nac!
